concurrency/ServiceController.java

```diff
@@ -0,0 +1,1922 @@
+        // and receivers
+        synchronized (sControllerLock) {
+            if (sServiceThread != null) {
+                mResolver = getContentResolver();
+
+                // Set up our observers; we need them to know when to start/stop various syncs based
+                // on the insert/delete/update of mailboxes and accounts
+                mMailboxObserver = new MailboxObserver(mHandler);
+                mResolver.registerContentObserver(Mailbox.SERVICE_CONTROLLER_NOTIFIER_URI, true,
+                    mMailboxObserver);
+
+                registerAccountObservers();
+
+                // We observe synced messages to trigger upsyncs at the appropriate time
+                mSyncedMessageObserver = new SyncedMessageObserver(mHandler);
+                mResolver.registerContentObserver(Message.NOTIFIER_URI, true,
+                    mSyncedMessageObserver);
+
+                // Set up receivers for connectivity and background data setting
+                mConnectivityReceiver = new ConnectivityReceiver();
+                registerReceiver(mConnectivityReceiver, new IntentFilter(
+                    ConnectivityManager.CONNECTIVITY_ACTION));
+
+                mBackgroundDataSettingReceiver = new ConnectivityReceiver();
+                registerReceiver(mBackgroundDataSettingReceiver, new IntentFilter(
+                    ConnectivityManager.ACTION_BACKGROUND_DATA_SETTING_CHANGED));
+                // Save away the current background data setting; we'll keep track of it with the
+                // receiver we just registered
+                mBackgroundData = sConnectivityManager.getBackgroundDataSetting();
+            }
+        }
+
+        try {
+            // Loop indefinitely until we're shut down
+            while (!sStop) {
+                try {
+                    sPowerHandler.runAwake(null);
+                    waitForConnectivity();
+                    mNextWaitReason = null;
+                    long nextWait = checkMailboxes();
+                    synchronized (sWaitLock) {
+                        if (!sKicked) {
+                            if (nextWait < 0) {
+                                Logging.log(TAG, "Negative wait? Setting to 1s");
+                                nextWait = MIN_WAIT_TIME;
+                            }
+                            if (nextWait > MIN_SLEEP_TIME) {
+                                if (mNextWaitReason != null) {
+                                    Logging.log(TAG, "Next awake ", nextWait / 1000, "s: ",
+                                            mNextWaitReason);
+                                }
+                                sPowerHandler.runAsleep(null, nextWait + SLEEP_EXTRA_TIME);
+                            }
+                            sWaitLock.wait(nextWait);
+                        }
+                    }
+                } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)